circle/src/cfft/mod.rs

```diff
@@ -1,24 +1,46 @@
 use alloc::vec;
 use alloc::vec::Vec;
 
-use itertools::{iterate, izip, Itertools};
-use p3_commit::PolynomialSpace;
-use p3_dft::{divide_by_height, Butterfly, DifButterfly, DitButterfly};
+use itertools::{iterate, Itertools};
 use p3_field::extension::ComplexExtendable;
-use p3_field::{batch_multiplicative_inverse, ExtensionField, Field};
+use p3_field::{ExtensionField, Field};
 use p3_matrix::dense::RowMajorMatrix;
 use p3_matrix::Matrix;
 use p3_maybe_rayon::prelude::*;
-use p3_util::{log2_ceil_usize, log2_strict_usize, reverse_slice_index_bits};
-use tracing::{debug_span, instrument};
+use p3_util::reverse_slice_index_bits;
+use tracing::instrument;
 
 use crate::domain::CircleDomain;
 use crate::point::{compute_lagrange_den_batched, Point};
 use crate::{cfft_permute_index, cfft_permute_slice, CfftPermutable, CfftView};
 
+mod par_chunked;
+
+pub use par_chunked::ParChunkedCfft;
+
+pub trait CfftAlgorithm<F: ComplexExtendable> {
+    fn interpolate<M: Matrix<F>>(&self, evals: CircleEvaluations<F, M>) -> RowMajorMatrix<F>;
+
+    fn evaluate(&self, domain: CircleDomain<F>, coeffs: RowMajorMatrix<F>) -> CircleEvaluations<F>;
+
+    #[instrument(skip_all, fields(
+        dims = %evals.values.dimensions(),
+        target = target_domain.log_n,
+    ))]
+    fn extrapolate<M: Matrix<F>>(
+        &self,
+        target_domain: CircleDomain<F>,
+        evals: CircleEvaluations<F, M>,
+    ) -> CircleEvaluations<F> {
+        assert!(target_domain.log_n >= evals.domain.log_n);
+        self.evaluate(target_domain, self.interpolate(evals))
+    }
```

Comment:
Could we add a comment here giving what the expected input/output will be? E.g. What column rearranging goes on before/after we plug into this?

Is the output bitreversed?

Reply:
Well, the whole point of `CircleEvaluations` is to hide the ordering. You have to call `evals.to_cfft_order()` or `evals.to_natural_order()` to observe the values

Reply:
Cool, that makes sense. I guess what I meant is, is CircleEvaluations being initialized using `from_natural_order`? Depending on the CFFT algorithm it might be useful to be able to initialize it in a different way.

E.g. have CircleEvaluations coded so that it starts in natural_order not cfft_order and then the CFFT algorithm can decide on what order to return it with.